From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Sat, 20 May 2017 11:32:17 +0200
Subject: [PATCH] {PRINTF} scripts/kconfig: add printf format attributes

When building with -Wformat-security and -Wsuggest-attribute=format in
HOSTCFLAGS, gcc complains about several functions misssing a printf
format attribute. Add them.
---
 scripts/kconfig/confdata.c        |  6 ++++--
 scripts/kconfig/lkc.h             |  6 ++++--
 scripts/kconfig/lxdialog/dialog.h |  4 ++--
 scripts/kconfig/menu.c            |  3 ++-
 scripts/kconfig/nconf.c           |  6 ++++--
 scripts/kconfig/zconf.y           | 12 ++++++++----
 6 files changed, 24 insertions(+), 13 deletions(-)

diff --git a/scripts/kconfig/confdata.c b/scripts/kconfig/confdata.c
index 91d0a5c014ac..429f39193828 100644
--- a/scripts/kconfig/confdata.c
+++ b/scripts/kconfig/confdata.c
@@ -90,7 +90,8 @@ static int conf_lineno, conf_warnings;
 
 const char conf_defname[] = "arch/$(ARCH)/defconfig";
 
-static void conf_warning(const char *fmt, ...)
+static __attribute__ ((format(printf, 1, 2)))
+void conf_warning(const char *fmt, ...)
 {
 	va_list ap;
 	va_start(ap, fmt);
@@ -115,7 +116,8 @@ void conf_set_message_callback(void (*fn)(const char *s))
 	conf_message_callback = fn;
 }
 
-static void conf_message(const char *fmt, ...)
+static void __attribute__ ((format(printf, 1, 2)))
+conf_message(const char *fmt, ...)
 {
 	va_list ap;
 	char buf[4096];
diff --git a/scripts/kconfig/lkc.h b/scripts/kconfig/lkc.h
index 9eb7c837cd8f..eb4f2f0b9998 100644
--- a/scripts/kconfig/lkc.h
+++ b/scripts/kconfig/lkc.h
@@ -82,7 +82,8 @@ static inline void xfwrite(const void *str, size_t len, size_t count, FILE *out)
 
 /* menu.c */
 void _menu_init(void);
-void menu_warn(struct menu *menu, const char *fmt, ...);
+void menu_warn(struct menu *menu, const char *fmt, ...)
+	__attribute__ ((format(printf, 2, 3)));
 struct menu *menu_add_menu(void);
 void menu_end_menu(void);
 void menu_add_entry(struct symbol *sym);
@@ -115,7 +116,8 @@ struct gstr {
 struct gstr str_new(void);
 void str_free(struct gstr *gs);
 void str_append(struct gstr *gs, const char *s);
-void str_printf(struct gstr *gs, const char *fmt, ...);
+void str_printf(struct gstr *gs, const char *fmt, ...)
+	__attribute__ ((format(printf, 2, 3)));
 const char *str_get(struct gstr *gs);
 
 /* symbol.c */
diff --git a/scripts/kconfig/lxdialog/dialog.h b/scripts/kconfig/lxdialog/dialog.h
index 0b00be5abaa6..865c57589792 100644
--- a/scripts/kconfig/lxdialog/dialog.h
+++ b/scripts/kconfig/lxdialog/dialog.h
@@ -152,8 +152,8 @@ extern int saved_x, saved_y;		/* Needed in signal handler in mconf.c */
 
 /* item list as used by checklist and menubox */
 void item_reset(void);
-void item_make(const char *fmt, ...);
-void item_add_str(const char *fmt, ...);
+__attribute__ ((format(printf, 1, 2))) void item_make(const char *fmt, ...);
+__attribute__ ((format(printf, 1, 2))) void item_add_str(const char *fmt, ...);
 void item_set_tag(char tag);
 void item_set_data(void *p);
 void item_set_selected(int val);
diff --git a/scripts/kconfig/menu.c b/scripts/kconfig/menu.c
index 4cf15d449c05..7166c95fd0fd 100644
--- a/scripts/kconfig/menu.c
+++ b/scripts/kconfig/menu.c
@@ -28,7 +28,8 @@ void menu_warn(struct menu *menu, const char *fmt, ...)
 	va_end(ap);
 }
 
-static void prop_warn(struct property *prop, const char *fmt, ...)
+static __attribute__ ((format(printf, 2, 3)))
+void prop_warn(struct property *prop, const char *fmt, ...)
 {
 	va_list ap;
 	va_start(ap, fmt);
diff --git a/scripts/kconfig/nconf.c b/scripts/kconfig/nconf.c
index 1ef232ae5ab9..8a194deb2ab3 100644
--- a/scripts/kconfig/nconf.c
+++ b/scripts/kconfig/nconf.c
@@ -522,7 +522,8 @@ static int get_mext_match(const char *match_str, match_f flag)
 }
 
 /* Make a new item. */
-static void item_make(struct menu *menu, char tag, const char *fmt, ...)
+static __attribute__ ((format(printf, 3, 4)))
+void item_make(struct menu *menu, char tag, const char *fmt, ...)
 {
 	va_list ap;
 
@@ -562,7 +563,8 @@ static void item_make(struct menu *menu, char tag, const char *fmt, ...)
 }
 
 /* very hackish. adds a string to the last item added */
-static void item_add_str(const char *fmt, ...)
+static __attribute__ ((format(printf, 1, 2)))
+void item_add_str(const char *fmt, ...)
 {
 	va_list ap;
 	int index = items_num-1;
diff --git a/scripts/kconfig/zconf.y b/scripts/kconfig/zconf.y
index 22fceb264cf5..a88cc23c6204 100644
--- a/scripts/kconfig/zconf.y
+++ b/scripts/kconfig/zconf.y
@@ -22,8 +22,10 @@ int cdebug = PRINTD;
 
 int yylex(void);
 static void yyerror(const char *err);
-static void zconfprint(const char *err, ...);
-static void zconf_error(const char *err, ...);
+static void zconfprint(const char *err, ...)
+	__attribute__ ((format(printf, 1, 2)));
+static void zconf_error(const char *err, ...)
+	__attribute__ ((format(printf, 1, 2)));
 static bool zconf_endtoken(const struct kconf_id *id, int starttoken, int endtoken);
 
 struct symbol *symbol_hash[SYMBOL_HASHSIZE];
@@ -598,7 +600,8 @@ static bool zconf_endtoken(const struct kconf_id *id, int starttoken, int endtok
 	return true;
 }
 
-static void zconfprint(const char *err, ...)
+static __attribute__ ((format(printf, 1, 2)))
+void zconfprint(const char *err, ...)
 {
 	va_list ap;
 
@@ -609,7 +612,8 @@ static void zconfprint(const char *err, ...)
 	fprintf(stderr, "\n");
 }
 
-static void zconf_error(const char *err, ...)
+static __attribute__ ((format(printf, 1, 2)))
+void zconf_error(const char *err, ...)
 {
 	va_list ap;
 
-- 
